src/Polly.Core/Timeout/OnTimeoutEvent.cs

```diff
@@ -0,0 +1,78 @@
+namespace Polly.Timeout;
+
+/// <summary>
+/// An event that is raised when a timeout occurs.
+/// </summary>
+/// <remarks>This class supports registering multiple on-timeout callbacks.</remarks>
+public sealed class OnTimeoutEvent
+{
+    private readonly List<Func<OnTimeoutArguments, ValueTask>> _events = new();
+
+    /// <summary>
+    /// Adds an asynchronous on-timeout callback.
+    /// </summary>
+    /// <param name="onTimeout">The specified callback.</param>
+    /// <returns>This instance.</returns>
+    public OnTimeoutEvent Add(Func<OnTimeoutArguments, ValueTask> onTimeout)
+    {
+        Guard.NotNull(onTimeout);
+
+        _events.Add(onTimeout);
+        return this;
+    }
+
+    /// <summary>
+    /// Adds a synchronous on-timeout callback.
+    /// </summary>
+    /// <param name="onTimeout">The specified callback.</param>
+    /// <returns>This instance.</returns>
+    public OnTimeoutEvent Add(Action<OnTimeoutArguments> onTimeout)
+    {
+        Guard.NotNull(onTimeout);
+
+        _events.Add(args =>
+        {
+            onTimeout(args);
+            return default;
+        });
+        return this;
+    }
+
+    /// <summary>
+    /// Adds a synchronous on-timeout callback.
+    /// </summary>
+    /// <param name="onTimeout">The specified callback.</param>
+    /// <returns>This instance.</returns>
+    public OnTimeoutEvent Add(Action onTimeout)
```

Comment:
There's no need/value for a similar Func overload?

```
public OnTimeoutEvent Add(Func<ValueTask> onTimeout)
```

Reply:
My thinking was that for async callbacks not to encourage too much. If someone want to use them, they should also consume the arguments.